### ISSUE TITLE
[Relay][RFC] Automatic Differentiation

### ISSUE THREAD

**Issue by MarisaKirisame** [dmlc/tvm#2237]
This RFC aims to pave the road to add automatic differentiation (AD) to Relay, both a first-order and a higher-order algorithm. An AD algorithm calculates the gradient of a program, which is needed for training (back propagation). Additionally, higher-order gradients are sometimes needed for other optimization algorithms such as Newton's method.

Because Relay support closures and control flow, and there are further plans to add features like algebraic data type (#2175) , our implementation of AD must support back propagation over these constructs. For the implementation, we plan to closely follow "[Demystifying Differentiable Programming](https://arxiv.org/pdf/1803.10228.pdf)." On a high level, our approach generates a graph at runtime and runs reverse-mode automatic differentiation on it as if it were a static graph.
This algorithm easily supports closures, control flow, and ADTs and it can be extended to account for further language features.

Differentiating programs that include operators requires us to register implementations of gradients for the different operators, which we will include as attributes. More specfically, for an operator `f : <x0, x1, x2...> -> y`,  the gradient of the operator is `f : <x0, x1, x2...> -> <y, y -> <x0, x1, x2...>>`
It's signature in C++ is runtime::TypedPackedFunc<Expr()>.
The signature for this attibute is open for discussion: we can scale other forms to the higher-order case as well.

First-order AD can be easily optimized and does not require any further language features beyond gradients for operators. AD on Higher-order program in the manner of demyst would require us to add OCaml-style references to Relay, which we plan to submit soon as a PR.

We would appreciate the community's feedback on this outline for implementing automatic differentiation in Relay. We welcome and would be glad to respond to any comments regarding further details about the implementation of AD and necessary steps to incorporate into Relay.
Thanks @slyubomirsky for helping me write this RFC.

**Comment by masahi** [comment]
Hi @MarisaKirisame, can you suggest some fast-path background readings to understand AD implementation in Relay? Besides the "Demystifying" paper?

**Comment by MarisaKirisame** [comment]
@masahi it is the fastest path.

It use delimited continuation so it is confusing, but it can be explained without.

Essentially, for every Double expression d, we transform it to an expression of type (Double, Ref Double).
It is a tuple which hold two value: the original value, and the gradient of that value.

There is also a global Ref (() -> ()) function (it take no argument, produce an empty tuple) called backward which is resposible for reading the gradient, and writing it upstream.

take expression `x + y` where `x` and `y` are subexpressions for example.
we will 
0: transform x and y into pair (x, xref) and (y, yref)
1: generate 
`
let zref = new 0;
let old_backward = !backward;
let new_backward = fun() { xref <- !xref + !zref; yref <- !yref + !zref; old_backward() };
backward <- new_backward; 
(x + y, zref)
`

There will be wrapper code which initialize a backward function which does nothing, and convert between Double and (Double, Ref Double)

it is essentially the same for tensor.

**Comment by MarisaKirisame** [comment]
@tqchen as per the request the signature in C++ side is runtime::TypedPackedFunc<Expr()>
I have updated it in above as well.

**Comment by tqchen** [comment]
@MarisaKirisame can you give a specific comment on how is this FGradient expected to be used?
```runtime::TypedPackedFunc<Expr()>``` does not really make sense because it does not take in input expressions. 

I was thinking more along the following line:

```c++
TypedPackedFunc<Array<Expr> (Expr orig_call, Expr out_grad)>
```
- Take in original call node, out_grad, and output, return list of gradient wrt to each input

**Comment by MarisaKirisame** [comment]
@tqchen it is a function expression taking orig_call, and will return a tuple. the right hand side will take out_grad. (it is a function in relay but not a function in C++).

**Comment by tqchen** [comment]
I see what you mean, it is possible to express simple function's gradient purely in relay. So conceptually, it is more like




 However, it could be hard to do so when some of the expression involves attributes like sum(x, axis=1).

Because we do not have a proper way to refer to the axis 


I see there are advantages of expressing the gradient function in relay. However

**Comment by MarisaKirisame** [comment]
@botev 
Higher order here mean AD in higher order language. It is indeed true that higher order AD of first order program need no further feature. We value the closure property (befor/after AD is in the same language) so for whatever AD algorithm you choose we can hopefully do it multiple time.

Sorry for the confusion, i had updated the RFC.

We had not thought about supporting forward mode as of right now, as our main priority is to get back propagation working.

**Comment by reminisce** [comment]
Maybe this is not in the scope of this RFC or extra work needs to be added, but can this feature be extended to support AD in imperative programming? NNVM is able to achieve that purpose and that's how it works in MXNet/Gluon APIs. Just want to make sure this design has taken imperative programming into consideration.

Eventually, can we achieve the UX as in https://github.com/HIPS/autograd?

**Comment by reminisce** [comment]
@MarisaKirisame Sorry for the confusion. Here is an introduction of [imperative and symbolic programming](https://mxnet.incubator.apache.org/architecture/program_model.html#case-study-backprop-and-autodiff) in MXNet.

Basically, imperative programming means the computation of operators are executed one by one so that users can still get intermediate layer outputs of a model. In the [AD PR](https://github.com/dmlc/tvm/pull/2321), it seems that current interface (Evaluator) optimizes both forward and backward graphs together so that users won't be able to get intermediate outputs of the forward graph.

The ideal use case would be when the program invokes operators one by one, a forward graph is constructed on the fly. When it starts to run backpropagation, it first constructs a backward graph and runs the backward graph independently. Any idea how to use Relay to realize this?

**Comment by MarisaKirisame** [comment]
@reminisce 
this is an orthgonal problem.
we are working on converting a define-by-run(imperative model) to a define-then-run(symbolic model)

**Comment by tqchen** [comment]
My understanding is that the current signature support imperative AD as well

**Comment by reminisce** [comment]
@MarisaKirisame Thanks. That's what I thought. No pressure to put it in this PR. Has an RFC been submitted?

@tqchen Could you elaborate your thoughts? My understanding is that we will need to attach node info to NDArray in the forward pass to construct Relay AST at runtime, and create a backward AST for running AD. Thanks.

**Comment by tqchen** [comment]
The current AD signature: ```Array<Expr> (Expr orig_call, Expr out_grad)``` constructs the gradient AST for a given input.  

Imperative AD can make use of this signature along with a high-level taping structure(possibly attached to NDArray of the framework) to get the gradient. So you don't need another separate signature for imperative AD.

**Comment by reminisce** [comment]
@tqchen Thanks for the answer. I will take a look.

**Comment by MarisaKirisame** [comment]
@reminisce not yet. it is still wip.

**Comment by MarisaKirisame** [comment]
AD is finished in #2496 .
